src/core/net/dns_headers.cpp

```diff
@@ -128,7 +128,7 @@ otError Name::AppendMultipleLabels(const char *aLabels, Message &aMessage)
 
     do
     {
-        VerifyOrExit(index < kMaxLength, error = OT_ERROR_INVALID_ARGS);
+        VerifyOrExit(index <= kMaxLength, error = OT_ERROR_INVALID_ARGS);
```

Comment:
I think the condition here `index < kMaxLength` is correct (in the sense that we want `index < 255` to be checked here and if `index` gets to `255` then the name should be considered as invalid).

My understanding is that the entire encoded name should be `255` bytes of less. Note that the name should always end with a "terminator" or a zero length byte. Here in `AppendMultipleLabels()` we should not go to 255 to leave room for the "terminator" zero byte (which is added by `AppendName()`).

That said, it may be good to clarify the definition of `kMaxLength` itself and what it represents.
- If we consider a string representation of a DNS name (i.e. a string of dot `'.'` seperated sequence of labels), then the string representation's length can be at most `254` chars (or `253` char if the string omits the the ending `'.'` , i.e., using `example.com` instead o`example.com.` ).
   -  I am using "string length" definition here that does not include the `'\0'` null char at the end of string which is the commonly used notion.
   - So if the string representation has a length of `255` then it is invalid.
- Here is reason for `254` length:
  - The encoded name format is sequence of `"len (byte)"` followed by `len` chars ending with `"zero len (byte)"` and is limited to 255 bytes total.
  - When we represent it as string we use dot `'.'` for all `"len (byte)"` except for the very first one, so we end up with `254` chars as max length.
  - Note that this assumes there would no escaping of chars in the name (which is possible/recommended in case of "service instance name" (which can contain more characters than other labels).

Here is my suggestion:
- We define a private `enum` value of `kMaxEncodedSize = 255` to represent max encoded size and and use in the `AppendMultipleLabels()` to make this clear.
- We can also clarify/update the definition of `Name::kMaxLength` that it represent the maximum length of string representation of a DNS name 
   - I actually have such change  in the DNS-SD branch that I am working on. In that we define `kMaxSize = 255` (size instead of length) to indicate the "size" of char array to pass to read a name (which is again `255` to include the null char at the end of string).
  

(see related discussion here [stackoverflow.com/what-is-the-maximum-length-of-a-dns-name](https://stackoverflow.com/questions/32290167/what-is-the-maximum-length-of-a-dns-name)).

Reply:
Yes, 255 is the max encoded size and the max length of a domain name string should be 254 if it ends with a dot, and 253 if it does not end with a dot. 

Yet, I think the original checking condition `index < kMaxLength` is not strictly correct. It might treat some invalid string as valid. 

I have refined the code accordingly so that it really check the encoded size against max length 255. 
Please have a look.